File: src/main/java/hk/ust/cse/comp4321/proj1/Crawler.java
package hk.ust.cse.comp4321.proj1;

import hk.ust.cse.comp4321.proj1.nlp.NLPUtils;
import org.jsoup.Connection;
import org.jsoup.Connection.Response;
import org.jsoup.HttpStatusException;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import javax.net.ssl.SSLHandshakeException;
import java.io.IOException;
import java.net.URL;
import java.util.*;
import java.util.stream.Collectors;

/**
 * The data structure for the crawling queue.
 */
class Link {
    URL url;
    int level;

    Link(URL url, int level) {
        this.url = url;
        this.level = level;
    }
}

/**
 * This is customized exception for those pages that have been visited before.
 */
class RevisitException
        extends RuntimeException {
    public RevisitException() {
        super();
    }
}

public class Crawler {
    private final URL rootURL;

    private final HashSet<URL> urls;     // the set of urls that have been visited before
    public Vector<Link> todos; // the queue of URLs to be crawled
    private int counter = 0; // to count the number of retrieved pages
    private final List<DocumentRecord> documentRecords = new ArrayList<>();

    private final int maxPages; // max page
    private final int maxDepth;

    Crawler(URL rootURL, int maxPages, int maxDepth) {
        this.rootURL = rootURL;
        this.todos = new Vector<>();
        this.todos.add(new Link(rootURL, 1));
        this.urls = new HashSet<>();
        this.maxPages = maxPages;
        this.maxDepth = maxDepth;
    }

    /**
     * Send an HTTP request and analyze the response.
     *
     * @return {Response} res
     * @throws HttpStatusException for non-existing pages
     * @throws IOException
     */
    public Response getResponse(URL url) throws HttpStatusException, IOException {
        if (this.urls.contains(url)) {
            throw new RevisitException(); // if the page has been visited, break the function
        }

        // the default body size is 2Mb, to attain unlimited page, use the following.
        // Connection conn = Jsoup.connect(this.url).maxBodySize(0).followRedirects(false);
        Connection conn = Jsoup.connect(url.toString()).followRedirects(false);

        /* establish the connection and retrieve the response */
        Response res = conn.execute();

        /* if the link redirects to other place... */
        if (res.hasHeader("location")) {
            URL actual_url = new URL(url, res.header("location"));
            if (this.urls.contains(actual_url)) {
                throw new RevisitException();
            } else {
                this.urls.add(actual_url);
            }
        } else {
            this.urls.add(url);
        }
        return res;
    }

    /**
     * Use a queue to manage crawl tasks.
     */
    public void crawlLoop() {
        while (!todos.isEmpty()) {
            Link focus = todos.remove(0);
            if (focus.level > maxDepth) break; // stop criteria
            if (urls.contains(focus.url)) continue;   // ignore pages that has been visited
            if (counter >= maxPages) {  // stop when number of pages checked out exceed the constant
                break;
            } else {
                counter++;
            }

            /* start to crawl on the page */
            try {
                Response returns = this.getResponse(focus.url);
                Response res = returns.bufferUp();
                Document doc = res.parse();

                // Check lang
                String htmlLang = doc.select("html").first().attr("lang");
                String bodyLang = doc.select("body").first().attr("lang");
                String lang = htmlLang + bodyLang;
                if (!lang.toLowerCase().contains("en")) {
                    System.out.printf("\n skipped link= %s\n", focus.url);
                    continue;
                }
                // Check lang end

                Vector<String> words = CrawlUtils.extractWords(doc).stream()
                        .filter(NLPUtils::isAlphaNumeric)
                        .filter(NLPUtils::stopwordFilter)
                        .map(NLPUtils::stemFilter)
                        .collect(Collectors.toCollection(Vector::new));

                Vector<String> extractedLinks = CrawlUtils.extractLinks(doc);
                Set<URL> linkSet = new HashSet<>();
                for (String extractedLink : extractedLinks) {
                    URL link = CrawlUtils.urlPreprocess(focus.url, extractedLink);
                    if (link.getHost().equals(rootURL.getHost())) {
                        todos.add(new Link(link, focus.level + 1)); // add links
                        linkSet.add(link);
                    }
                }
                ArrayList<URL> linksList = new ArrayList<>(linkSet);

                // retrieving data
                String lastModified = res.header("Last-Modified");
                if (lastModified == null || lastModified.equals("")) {
                    lastModified = res.header("Date");
                }
                int size = res.bodyAsBytes().length;

                // count keywords
                TreeMap<String, Integer> freqTable = new TreeMap<>();
                for (String item : words) {
                    freqTable.put(item, freqTable.getOrDefault(item, 0) + 1);
                }

                // Calling document record to serialise the retrieved data
                DocumentRecord documentRecord = new DocumentRecord(focus.url);
                documentRecord.setTitle(res.parse().title());
                documentRecord.setLastModificationDate(new Date(lastModified));
                documentRecord.setFreqTable(freqTable);
                documentRecord.setPageSize(size);
                documentRecord.setChildLinks(linksList);
                documentRecord.setWords(words);

                documentRecords.add(documentRecord);

            } catch (HttpStatusException e) {
                // e.printStackTrace ();
                System.out.printf("\nLink Error: %s\n", focus.url);
            } catch (SSLHandshakeException e) {
                System.out.printf("\nSSL Error: %s\n", focus.url);
            } catch (IOException e) {
                e.printStackTrace();
            } catch (RevisitException ignored) {
            } catch (Exception e) {
                System.out.printf("\nUnhandled error: %s\n", e.getMessage());
            }
        }

    }


    public List<DocumentRecord> getDocumentRecords() {
        return documentRecords;
    }

}

File: src/main/kotlin/hk/ust/cse/comp4321/proj1/Crawler.kt
package hk.ust.cse.comp4321.proj1

import hk.ust.cse.comp4321.proj1.nlp.NLPUtils
import org.jsoup.Connection
import org.jsoup.HttpStatusException
import org.jsoup.Jsoup
import java.io.IOException
import java.net.URL
import java.util.*
import javax.net.ssl.SSLHandshakeException


class Crawler constructor(private val rootURL: URL, private val maxPages: Int, private val maxDepth: Int) {
    /**
     * The data structure for the crawling queue.
     */
    class Link(var url: URL, var level: Int)

    /**
     * This is customized exception for those pages that have been visited before.
     */
    class RevisitException : RuntimeException()

    private val urls // the set of urls that have been visited before
            : HashSet<URL> = HashSet()
    var todos // the queue of URLs to be crawled
            : Vector<Link> = Vector()
    private var counter = 0 // to count the number of retrieved pages
    val documentRecords: MutableList<DocumentRecord> = ArrayList()

    /**
     * Send an HTTP request and analyze the response.
     *
     * @return {Response} res
     * @throws HttpStatusException for non-existing pages
     * @throws IOException
     */
    @Throws(HttpStatusException::class, IOException::class)
    fun getResponse(url: URL): Connection.Response {
        if (urls.contains(url)) {
            throw RevisitException() // if the page has been visited, break the function
        }

        // the default body size is 2Mb, to attain unlimited page, use the following.
        // Connection conn = Jsoup.connect(this.url).maxBodySize(0).followRedirects(false);
        val conn = Jsoup.connect(url.toString()).followRedirects(false)

        /* establish the connection and retrieve the response */
        val res = conn.execute()

        /* if the link redirects to other place... */if (res.hasHeader("location")) {
            val actualUrl = URL(url, res.header("location"))
            if (urls.contains(actualUrl)) {
                throw RevisitException()
            } else {
                urls.add(actualUrl)
            }
        } else {
            urls.add(url)
        }
        return res
    }

    /**
     * Use a queue to manage crawl tasks.
     */
    fun crawlLoop() {
        while (!todos.isEmpty()) {
            val focus = todos.removeAt(0)
            if (focus.level > maxDepth) break // stop criteria
            if (urls.contains(focus.url)) continue  // ignore pages that has been visited
            if (counter >= maxPages) {  // stop when number of pages checked out exceed the constant
                break
            } else {
                counter++
            }

            /* start to crawl on the page */try {
                val returns = getResponse(focus.url)
                val res = returns.bufferUp()
                val doc = res.parse()

                // Check lang
                val htmlLang = doc.select("html").first()!!.attr("lang")
                val bodyLang = doc.select("body").first()!!.attr("lang")
                val lang = htmlLang + bodyLang
                if (!lang.lowercase(Locale.getDefault()).contains("en")) {
                    System.out.printf("\n skipped link= %s\n", focus.url)
                    continue
                }
                // Check lang end
                val words = CrawlUtils.extractWords(doc)
                        .filter { NLPUtils.isAlphaNumeric(it) }
                        .filter { NLPUtils.stopwordFilter(it) }
                        .map { NLPUtils.stemFilter(it) }
                val extractedLinks = CrawlUtils.extractLinks(doc)
                val linkSet: MutableSet<URL> = HashSet()
                for (extractedLink in extractedLinks) {
                    val link = CrawlUtils.urlPreprocess(focus.url, extractedLink)
                    if (link.host == rootURL.host) {
                        todos.add(Link(link, focus.level + 1)) // add links
                        linkSet.add(link)
                    }
                }
                val linksList = ArrayList(linkSet)

                // retrieving data
                var lastModified = res.header("Last-Modified")
                if (lastModified == null || lastModified == "") {
                    lastModified = res.header("Date")
                }
                val size = res.bodyAsBytes().size

                // count keywords
                val freqTable = TreeMap<String, Int>()
                for (item in words) {
                    freqTable[item] = freqTable.getOrDefault(item, 0) + 1
                }

                // count title as keywords with higher weight,
                val title = res.parse().title()
                val titleFreqTable = TreeMap<String, Int>()
                val tokenizedTitle = CrawlUtils.extractTitleWords(title)
                        .filter { NLPUtils.isAlphaNumeric(it) }
                        .filter { NLPUtils.stopwordFilter(it) }
                        .map { NLPUtils.stemFilter(it) }
                for (item in tokenizedTitle) {
                    titleFreqTable[item] = titleFreqTable.getOrDefault(item, 0) + 1
                }

                // Calling document record to serialise the retrieved data
                val documentRecord = DocumentRecord(focus.url)
                documentRecord.setTitle(title)
                        .setLastModificationDate(Date(lastModified))
                        .setFreqTable(freqTable)
                        .setTitleFreqTable(titleFreqTable)
                        .setPageSize(size)
                        .setChildLinks(linksList)
                        .setWords(words)
                        .setTitleWords(tokenizedTitle)
                documentRecords.add(documentRecord)
            } catch (e: HttpStatusException) {
                // e.printStackTrace ();
                System.out.printf("\nLink Error: %s\n", focus.url)
            } catch (e: SSLHandshakeException) {
                System.out.printf("\nSSL Error: %s\n", focus.url)
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (ignored: RevisitException) {
            } catch (e: Exception) {
                System.out.printf("\nUnhandled error: %s\n", e.message)
            }
        }
    }

    init {
        todos.add(Link(rootURL, 1))
    }
}
